### ISSUE TITLE
Should we relax ONNX (checker) requirement that the model output type/shape should be specified?

### ISSUE THREAD

**Issue by gramalingam** [onnx/onnx#4934]
ONNX requires that the types, including shape, be specified for the inputs and outputs of a model.

This is a suggestion that this requirement be relaxed for *outputs*. Specifically, that would mean that the ONNX Checker does not complain if this requirement is violated.

The rationale for this suggestion is the following:

(a) For a proper model, ONNX's type-inference can infer the output-types of the model. So, there is no need to enforce this requirement. In practice, we have situations where an exporter uses ONNX's inference to fill in the output-types just so this requirement is met. We may as well make it official.

(b) However, ONNX's shape-inference is done on a best-effort basis. So, in principle, ONNX's inference may not be able to infer the shapes of the outputs. (For example, consider a Reshape node where the output-shape is a runtime input.) The current requirement, however, forces the model-creator to specify the output tensor's rank at least.

So, we do have this limitation to what inference can do. But ONNX already supports dynamic-shapes, in general. So, I think the above relaxation is a reasonable choice.

In other words: I think that the real issue some backends (especially based on compilation) care about is whether the shapes of all tensors (inputs, intermediates, and outputs) are fully known statically (which enables a number of optimizations) or not. The existing constraint, however, is not strong enough to guarantee that all shapes are statically known. So, again, the relaxation mentioned above may be reasonable to do.

Any thoughts?

**Comment by justinchuby** [comment]
I second this, for the reasons above

**Comment by xadupre** [comment]
I'm in favor, same reasons.

**Comment by fdwr** [comment]
@gramalingam I think this would be useful for general models that could accept either float32 or float16, likely needing multiple [`CastLike`](https://github.com/onnx/onnx/blob/main/docs/Operators.md#CastLike) calls in between the weights and the rest of the graph, which could be optimized away after model loading and before execution.

It would certainly be valuable to me. Currently when I slice an .onnx model apart into smaller fragments for easier debugging, I have to dynamically patch each model fragment's in-memory protobuf with the matching data types of inputs (using the output data type from the earlier fragment) before executing them with ONNX Runtime, because otherwise it rejects the model. Note ORT already accepts completely unspecified shapes (free dimension sizes and free dimension count).

If we relax this, ONNX implementations should also have a means to specify the data type early so that (like with constant folding and shape inference) they can optimize the model between model loading and execution. e.g. ONNX Runtime has a [`AddFreeDimensionOverrideByName`](https://github.com/microsoft/onnxruntime/blob/acc2ac627fc7948552aefc178e447c7860c04351/include/onnxruntime/core/session/onnxruntime_c_api.h#L2077) function.

1. *Model load*: ONNX implementation loads the model file (no type inference, shape inference, or optimizations done yet)
2. *Model query*: Application can query general properties of the loaded model like how many inputs/outputs it has and a list of acceptable data types. Unfortunately this valuable step is missing from some popular ONNX implementations (you can query the number inputs and their constraints *after* you load the model, but by then it's already too late to specify explicit values for static layout because optimizations were prematurely applied).
3. *Model binding/constraints*: Application specifies shapes and types either by name, or by binding inputs/output to it of that shape and type.
4. *Model optimization*: ONNX implementation _now_ runs shape inference, type inference, constant folding, execution provider allocation, partitioning, other optimizations (e.g. fusion)...
5. *Model execution*: Application begins execution of the now-optimized model via the ONNX implementation.

**Comment by cbourjau** [comment]
Relaxing the requirement may break quite a few legitimate use cases that are working today. Please don't do it! Here are a few points in no particular order why I think the current behavior is largely good:
 
#### Precedence in strongly typed languages
It is pretty common for strongly typed languages with type inference to require the user to spell out the concrete types at the function boundary, even if type inference could be used to precisely infer the output types. This is very useful for maintaining a stable interface and makes everything easier to comprehend by keeping the information more localized.

#### A third party cannot run type inference for custom operators
Custom operators usually won't have explicit libraries to run type and shape inference on them. The best guess is that whoever added the node to a graph in the first place knew the types and shape rules and set them correctly in the outputs of the graph. No third-party tool would be able to run type inference on a graph with custom operators.

##### Use case: Embedding graphs in other graphs 
Let's assume you write a library for constructing ONNX graphs with eager type inference. That library should be able to "wrap" an existing graph and inline it in the new graph. If the graph has its input and output types defined, it can be easily inlined without breaking the type inference even if the inlined graph contains custom operators.

#### Dealing with dynamically shaped tensors
The standard states:
> Each main (top-level) graph MUST define the names, types and shapes of its inputs and outputs, which are specified as ‘value info’ structures. The main graph inputs and outputs are required to have a shape, indicating the rank, even though the exact dimensions need not be specified.

I think it is important to keep in mind that the current standard only requires the rank, not the exact shape. While it probably is a bad idea to have a dynamic output rank it seems a bit odd that it is currently forbidden. If something needs to be changed in the standard, maybe this would be a minimally invasive point. Instead of throwing out the entire requirement of defining output types, we could allow the use of undefined shapes in the `ValueInfo`s of the outputs?

**Comment by gramalingam** [comment]
@cbourjau 

* The suggestion is to make output-types optional. Users can still specify them, and if they do, type-inference will verify that inferred type matches specified type.
* In practice, the ideal way forward is likely to make the behavior of the checker controllable via a user-option. Does that sound reasonable? If so, I guess a second question would be what the default behavior should be.
* I am not convinced about the custom-op point. What would the runtime do in the case of a custom-op where it doesn't have a type-inference method? I think ONNX's own inference will likely produce an error (I need to check this, just my vague recollection at this point) in this case, and a backend wouldn't be able to do any static type-checking anyway (that the custom-op is being called with correct type arguments).
* This does not prevent any frontend model-creators from doing what they do already ... if they create models with complete type information, they can still do so.

**Comment by cbourjau** [comment]
Let's be explicit. As I understand it, the proposal would be to change the standard to read something like this:
> Each main (top-level) graph MUST define the names, types and shapes of its inputs ~~and outputs~~, which are specified as ‘value info’ structures. The main graph inputs ~~and outputs~~ are required to have a shape, indicating the rank, even though the exact dimensions need not be specified.

This means that outputs, just like any other internal value, `MAY` have a `value info` associated with them. Use cases that rely on the output `value info`s are guaranteed to work today. After this change, they `MAY` work - or not. That is a big change.

Let's take a step back. I think there are (at least) three overlapping topics at hand here:
1. Type and shape information may have independent use cases. A tensor with a type of unknown rank is still more useful than an entirely untyped "value".
2. Users may have to perform an extra step after creating a graph to fill in the outputs' `value info`s
3. The standard allows tensors of dynamic rank, but the outputs must have static rank

About each point: 1. is more of an observation. 2.  can be greatly improved by better documentation. If type/shape inference fails, the person creating the model is still in a better position than anybody else to spell out the output `value info`s manually. 3. Should be answered on its own merits regardless of the current user experience of creating graphs with the `onnx` library today.

About your points in detail, @gramalingam :
> In practice, the ideal way forward is likely to make the behavior of the checker controllable via a user-option. Does that sound reasonable? If so, I guess a second question would be what the default behavior should be.

I think the checker should always precisely follow the standard. However, it should be obvious and easy for the user of `onnx` how to make their model standard compliant. 

> I am not convinced about the custom-op point. What would the runtime do in the case of a custom-op where it doesn't have a type-inference method?

For `onnx`, a custom operator is just another operator with no type/shape inference, isn't it? We even have operators from the standard where that is the case. At runtime, the user needs to provide the implementation of that custom operator and the particular runtime at hand. The `onnxruntime` for example requires the user to [explicitly](https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/session/onnxruntime_c_api.h#L4009) spell out the input and output types of a custom operation. There is no need for the author of a custom operator to provide anything that would hook into `onnx`'s shape/type inference. 

> This does not prevent any frontend model-creators from doing what they do already ... if they create models with complete type information, they can still do so.

But what about third-party services that have no control over the model creation? Changing the standard has a huge influence on their user's experience.

**Comment by gramalingam** [comment]
@cbourjau I agree that there are several different aspects here.

As a basic change, we could do something a little bit beyond what you suggest for (2). We could provide a utility that does the extra-step for the user: essentially something that performs what the checker + type/shape inference (checker) do, inferring the output type/rank and filling it in for the output, and failing if it was unable to produce the output type/rank. Further, this utility is best done by the existing checker/inference utilities. (We could add an option if we want to let the user decide whether this should be done or not.) This does not change the standard per se.

As for the rest: this is more an exploratory question to understand whether this existing constraint serves any useful purpose. I completely understand that existing code may assume that this constraint is satisfied. Hence, changing it may break some existing code. So, I do understand the value of retaining the constraint for compatibility reasons. But that is not the same as saying that there is a valid use-case for this constraint ... I am not sure if I am making myself clear. So, if we consider, say, ONNX 2.0, would we want this constraint or perhaps something else?

**Comment by jbachurski** [comment]
I think a lot of arguments have already been said here and I don't want to reiterate. Even though shape inference is 'best-effort', I think ONNX should strive to make it fully complete with as few exceptions possible and less 'best-effort'. The best way towards this goal is however a difficult research problem (and a huge amount of work, which can be made easier by reducing the number of ops).

Similarly, making models more 'polymorphic' (allowing more input element types) really mandates better support than relaxing constraints. Nevertheless, I could agree it would be nice to make utilities as flexible as possible to allow interesting uses of ONNX. Spox makes some use of such utilities (running shape inference on models without output types, which is almost undocumented).

However, I think the standard should stay with the statement that a truly valid model has a fully known signature. Even if ONNX Runtime (which has a step-by-step interpreter execution model) doesn't require all the type information, it would be nice for strongly typed languages to have a guarantee on what gets returned (and as such a nice signature can be created without fear something else will be returned).

**Comment by cbourjau** [comment]
@gramalingam I think I get your point about this being explorative :). Below are arguments which I think equally apply to "ONNX 2.0".

Given that the question (as I understand it) is about the type and rank there are three reasonable scenarios (I don't think anybody is advocating for "undefined type but mandatory rank"?). Below are some impacts/benefits as I see them:

**TLDR;** I believe that the restriction helps foster the ecosystem of third-party tools while not really posing much of a downside. 

#### Keep things as they are (type and rank of outputs are mandatory)

As stated earlier, the person making the model is in the best position to define the type and rank of the outputs. With good tooling this should usually be a oneliner.

Having the guarantee that every ONNX model has this information is important for third-party tools. For example, it allows a third-party tool to efficiently and easily compose models with strong static checks (a huge selling point of ONNX in my opinion) even if custom operators are involved.

While these tools could still exist if the annotations were optional, implementations tend to gravitate to the lowest common denominator of a standard: The calculus of whether to invest in such tools obviously shifts if the "total addressable market" goes from 100% to something unknown, probably marginal. If the tools don't exist that in turn means most libraries that create models don't have much of an incentive to implement the annotation of outputs in the first place.

#### Only the type is mandatory, the rank may be undefined

This would allow people to return arbitrarily ranked tensors at the expense of lessening the usefulness of static checks when composing models. 

I have a bit of a hard time seeing the use-case for returning arbitrarily ranked tensors, though. Also, if you really must, you could just flatten them and return the shape as an individual output within the currently existing restriction. Thus, I don't see how this scenario would open up new use cases but it would diminish what third-party tools can do while statically guaranteeing that your model works as expected.

#### Models are not required to annotate outputs at all

This would make it easier to implement a library that creates ONNX models because one would probably just not implement shape inference. This might lower the entry bar for new implementations. On the other hand, `onnx/onnx` already exists and it offers type and shape inference for the vast majority of the standard. Other projects should be able to easily use the tools provided by this project for their type and shape inference needs.
